File: src/capuchin_crypt_ffi.mjs
import { Result$Error, Result$Ok } from "./gleam.mjs";

const crypt = new Map();

export function crypt_erase(key) {
  if (crypt.has(key)) {
    crypt.delete(key);
    return true;
  }

  return false;
}

export function crypt_get(key) {
  if (crypt.has(key)) {
    return Result$Ok(crypt.get(key));
  }

  return Result$Error(undefined);
}

export function crypt_put(key, value) {
  crypt.set(key, value);
  return value;
}

File: src/capuchin_crypt_ffi.erl
-module(capuchin_crypt_ffi).
-export([crypt_erase/1, crypt_get/1, crypt_put/2]).

crypt_erase(Key) ->
    persistent_term:erase({capuchin_crypt, Key}).

crypt_get(Key) ->
    try persistent_term:get({capuchin_crypt, Key}) of
        Val -> {ok, Val}
    catch
        error:badarg -> {error, nil}
    end.

crypt_put(Key, Val) ->
    persistent_term:put({capuchin_crypt, Key}, Val),
    persistent_term:get({capuchin_crypt, Key}).
